laghos@3.1 build fails w/ mpich: laghos.cpp:87:4: error: reference to 'MPI_Session' is ambiguous

### Steps to reproduce the issue

`laghos@3.1` build is failing using:
* Ubuntu 20.04, AMD64
* GCC 11.4.0

Spack environment: [spack.yaml.txt](https://github.com/spack/spack/files/14252630/spack.yaml.txt)


<details><summary>Concretization</summary>

<pre>
 -   laghos@3.1%gcc@11.4.0+metis~ofast build_system=makefile arch=linux-ubuntu20.04-x86_64
[+]      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu20.04-x86_64
[+]      ^mfem@4.6.0%gcc@11.4.0~amgx~conduit~cuda~debug~examples~exceptions~fms~ginkgo~gnutls~gslib~hiop~lapack~libceed~libunwind+metis~miniapps~mpfr+mpi~netcdf~occa~openmp~petsc~pumi~raja~rocm~shared~slepc+static~strumpack~suite-sparse~sundials~superlu-dist~threadsafe~umpire+zlib build_system=generic patches=718f073 timer=auto arch=linux-ubuntu20.04-x86_64
[+]          ^hypre@2.30.0%gcc@11.4.0~caliper~complex~cuda~debug+fortran~gptune~int64~internal-superlu~magma~mixedint+mpi~openmp~rocm+shared~superlu-dist~sycl~umpire~unified-memory build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^openblas@0.3.26%gcc@11.4.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-ubuntu20.04-x86_64
[+]                  ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu20.04-x86_64
[+]                      ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64
[+]                      ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64
[+]                          ^diffutils@3.9%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]                      ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                          ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64
[+]          ^metis@5.1.0%gcc@11.4.0~gdb~int64~ipo~real64+shared build_system=cmake build_type=Release generator=make patches=4991da9,93a7903,b1225da arch=linux-ubuntu20.04-x86_64
[+]              ^cmake@3.27.9%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64
[+]                  ^curl@8.4.0%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-x86_64
[+]                      ^nghttp2@1.57.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                      ^openssl@3.1.3%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64
[+]                          ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]                      ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                  ^ncurses@6.3%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64
[e]          ^mpich@4.1.2%gcc@11.4.0~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64
[+]          ^zlib-ng@2.1.5%gcc@11.4.0+compat+opt build_system=autotools arch=linux-ubuntu20.04-x86_64
</pre></details>

Reproduced here using Docker container image `ecpe4s/ubuntu20.04-runner-amd64-gcc-11.4-mpi-base:2023.08.20` and publicly cached binaries:

```
$> docker run -it ecpe4s/ubuntu20.04-runner-amd64-gcc-11.4-mpi-base:2023.08.20
root@8f846a385d35:/# git clone https://github.com/spack/spack
root@8f846a385d35:/# git -C /spack checkout 2064934ed3c02edeab05bbd5719df6acf5743784
root@8f846a385d35:/# . spack/share/spack/setup-env.sh

root@8f846a385d35:/# spack mirror add E4S https://cache.e4s.io/24.02
root@8f846a385d35:/# spack buildcache keys -it

# Download spack.yaml attached near top of this issue description
root@8f846a385d35:/# spack -e . install
...
==> Installing laghos-3.1-heg3kkc2jwk7h7mayhiigsywak4b7w2b [9/9]
==> No binary for laghos-3.1-heg3kkc2jwk7h7mayhiigsywak4b7w2b found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/49/49b65edcbf9732c7f6c228958620e18980c43ad8381315a8ba9957ecb7534cd5.tar.gz
==> No patches needed for laghos
==> laghos: Executing phase: 'edit'
==> laghos: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'MFEM_DIR=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh' 'CONFIG_MK=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/share/mfem/config.mk' 'TEST_MK=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/share/mfem/test.mk'

3 errors found in build log:
     2     ==> laghos: Executing phase: 'build'
     3     ==> [2024-02-12-15:44:48.327674] 'make' '-j16' 'MFEM_DIR=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh' 'CONFIG_MK=/spack/
           opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/share/mfem/config.mk' 'TEST_MK=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0
           /mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/share/mfem/test.mk'
     4     cd .; /usr/local/mpich/install/mpich/bin/mpic++ -O3 -std=c++11 -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include -I/
           spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/hypre-2.30.0-tlpkuqe7cklavodwkaj3yclqzqfgqjsy/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/metis-5.1.0-
           wkqolmovas7we2lueutxa6ntyuik2gf7/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/zlib-ng-2.1.5-5x6i3wn4nd3uvng3wggbb6n7jjqrqpv6/include -I/spack/opt/spack/lin
           ux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem -c laghos.cpp
     5     cd .; /usr/local/mpich/install/mpich/bin/mpic++ -O3 -std=c++11 -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include -I/
           spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/hypre-2.30.0-tlpkuqe7cklavodwkaj3yclqzqfgqjsy/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/metis-5.1.0-
           wkqolmovas7we2lueutxa6ntyuik2gf7/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/zlib-ng-2.1.5-5x6i3wn4nd3uvng3wggbb6n7jjqrqpv6/include -I/spack/opt/spack/lin
           ux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem -c laghos_assembly.cpp
     6     cd .; /usr/local/mpich/install/mpich/bin/mpic++ -O3 -std=c++11 -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include -I/
           spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/hypre-2.30.0-tlpkuqe7cklavodwkaj3yclqzqfgqjsy/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/metis-5.1.0-
           wkqolmovas7we2lueutxa6ntyuik2gf7/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/zlib-ng-2.1.5-5x6i3wn4nd3uvng3wggbb6n7jjqrqpv6/include -I/spack/opt/spack/lin
           ux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem -c laghos_solver.cpp
     7     laghos.cpp: In function 'int main(int, char**)':
  >> 8     laghos.cpp:87:4: error: reference to 'MPI_Session' is ambiguous
     9        87 |    MPI_Session mpi(argc, argv);
     10          |    ^~~~~~~~~~~
     11    In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem/mfem.hpp:41,
     12                     from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem.hpp:2,
     13                     from laghos_solver.hpp:20,
     14                     from laghos.cpp:65:

     ...

     21                     from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem.hpp:2,
     22                     from laghos_solver.hpp:20,
     23                     from laghos.cpp:65:
     24    /usr/local/mpich/install/mpich/include/mpi.h:292:13: note:                 'typedef int MPI_Session'
     25      292 | typedef int MPI_Session;
     26          |             ^~~~~~~~~~~
  >> 27    laghos.cpp:88:21: error: 'mpi' was not declared in this scope
     28       88 |    const int myid = mpi.WorldRank();
     29          |                     ^~~
     30    laghos.cpp:220:27: warning: 'mfem::Mesh::Mesh(int, double)' is deprecated [-Wdeprecated-declarations]
     31      220 |          mesh = new Mesh(2);
     32          |                           ^
     33    In file included from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem/mesh/mesh_headers.hpp:26,

     ...

     57                     from /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem.hpp:2,
     58                     from laghos_solver.hpp:20,
     59                     from laghos.cpp:65:
     60    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/mfem-4.6.0-b7q3ztqv27llozzw73k6fiw4c2civosh/include/mfem/mesh/mesh.hpp:985:4: note: declared here
     61      985 |    Mesh(int nx, int ny, int nz, Element::Type type, bool generate_edges = false,
     62          |    ^~~~
  >> 63    make: *** [makefile:109: laghos.o] Error 1
     64    make: *** Waiting for unfinished jobs....
```

### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.22.0.dev0 (2064934ed3c02edeab05bbd5719df6acf5743784)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/14252652/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/14252653/spack-build-out.txt)

@tzanio @v-dobrev @vladotomov @sameershende 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/laghos/package.py
+++ b/var/spack/repos/builtin/packages/laghos/package.py
@@ -44,6 +44,13 @@ class Laghos(MakefilePackage):
     # Recommended mfem version for laghos v1.x is: ^mfem@3.3.1-laghos-v1.0
     depends_on("mfem@3.3.1-laghos-v1.0", when="@1.0,1.1")
 
+    # Replace MPI_Session
+    patch(
+        "https://github.com/CEED/Laghos/commit/c800883ab2741c8c3b99486e7d8ddd8e53a7cb95.patch?full_index=1",
+        sha256="e783a71c3cb36886eb539c0f7ac622883ed5caf7ccae597d545d48eaf051d15d",
+        when="@3.1 ^mfem@4.4:",
+    )
+
     @property
     def build_targets(self):
         targets = []

